File: ruby/lib/phonenumber_ae/native.rb
# frozen_string_literal: true

# Fiddle bindings for the phonenumber engine (libphonenumber_ae.so), ABI v5.
#
# This file is the ONLY place in the Ruby binding that knows about the C ABI.
# Everything above it (`phone_number.rb`) is idiomatic Ruby over these symbols.
# No phone-number logic lives here or anywhere else in this gem — the engine is
# `core/phonenumber.ae`, shared by every language binding.
#
# Library resolution, in order:
#   1. an explicit path passed to `Native.load(path)`
#   2. $LIBPHONENUMBER_AE_LIB       (what the in-tree .tests.ae leaves set)
#   3. native/ bundled next to this gem's lib dir (what a built gem ships)
#   4. the OS loader's own search path

require "fiddle"
require "rbconfig"

module PhoneNumberAe
  # Low-level FFI seam. Everything here mirrors core/embed.ae one-for-one.
  module Native
    LIB_NAME =
      case RbConfig::CONFIG["host_os"]
      when /darwin/ then "libphonenumber_ae.dylib"
      when /mswin|mingw|cygwin/ then "phonenumber_ae.dll"
      else "libphonenumber_ae.so"
      end

    # ---- format styles (ABI constants — append only, never renumber) ----
    E164 = 0
    INTERNATIONAL = 1
    NATIONAL = 2
    RFC3966 = 3

    # ---- number types (number_type result; -1 = unknown) ----
    TYPE_UNKNOWN = -1
    TYPE_FIXED_LINE = 0
    TYPE_MOBILE = 1
    TYPE_TOLL_FREE = 2
    TYPE_PREMIUM_RATE = 3
    TYPE_SHARED_COST = 4
    TYPE_VOIP = 5
    TYPE_PERSONAL_NUMBER = 6
    TYPE_PAGER = 7
    TYPE_UAN = 8
    TYPE_VOICEMAIL = 9
    TYPE_FIXED_LINE_OR_MOBILE = 10

    # ---- ValidationResult (is_possible_number_with_reason) ----
    VR_IS_POSSIBLE = 0
    VR_IS_POSSIBLE_LOCAL_ONLY = 4
    VR_INVALID_COUNTRY_CODE = 1
    VR_TOO_SHORT = 2
    VR_INVALID_LENGTH = 5
    VR_TOO_LONG = 3

    # ---- MatchType (is_number_match) ----
    MATCH_NOT_A_NUMBER = 0
    MATCH_NO_MATCH = 1
    MATCH_SHORT_NSN = 2
    MATCH_NSN = 3
    MATCH_EXACT = 4

    # ---- CountryCodeSource (pn_source) ----
    SRC_FROM_NUMBER_WITH_PLUS = 1
    SRC_FROM_NUMBER_WITH_IDD = 5
    SRC_FROM_NUMBER_WITHOUT_PLUS = 10
    SRC_FROM_DEFAULT_COUNTRY = 20

    # ---- matcher leniency ----
    LENIENCY_POSSIBLE = 0
    LENIENCY_VALID = 1

    # ---- ShortNumberCost (short_expected_cost) ----
    COST_TOLL_FREE = 0
    COST_STANDARD_RATE = 1
    COST_PREMIUM_RATE = 2
    COST_UNKNOWN = 3

    P = Fiddle::TYPE_VOIDP
    I = Fiddle::TYPE_INT
    V = Fiddle::TYPE_VOID

    # name => [argtypes, restype].
    #
    # Every string-returning symbol is declared as TYPE_VOIDP, not
    # Fiddle::TYPE_CONST_STRING: the pointer is caller-owned and has to come
    # back through aether_pn_embed_free_string. Letting Fiddle turn it into a
    # Ruby String directly would lose the pointer and leak the buffer.
    SIGS = {
      # lifecycle / metadata
      "aether_pn_embed_abi_version" => [[], I],
      "aether_pn_embed_free_string" => [[P], V],
      "aether_pn_embed_country_code" => [[P], P],
      "aether_pn_embed_example_number" => [[P], P],
      "aether_pn_embed_example_number_for_type" => [[P, I], P],
      "aether_pn_embed_invalid_example_number" => [[P], P],
      "aether_pn_embed_possible_lengths" => [[P], P],
      "aether_pn_embed_region_code_for_country_code" => [[P], P],
      "aether_pn_embed_is_nanpa_country" => [[P], I],
      "aether_pn_embed_ndd_prefix_for_region" => [[P, I], P],
      "aether_pn_embed_region_count" => [[], I],
      "aether_pn_embed_region_at" => [[I], P],
      "aether_pn_embed_cc_region_count" => [[P], I],
      "aether_pn_embed_cc_region_at" => [[P, I], P],
      # parse + parsed-number accessors
      "aether_pn_embed_parse" => [[P, P], P],
      "aether_pn_embed_national_number" => [[P, P], P],
      "aether_pn_embed_pn_region" => [[P], P],
      "aether_pn_embed_pn_country_code" => [[P], P],
      "aether_pn_embed_pn_national_number" => [[P], P],
      "aether_pn_embed_pn_extension" => [[P], P],
      "aether_pn_embed_pn_italian_leading_zero" => [[P], I],
      "aether_pn_embed_pn_source" => [[P], I],
      "aether_pn_embed_pn_error" => [[P], P],
      "aether_pn_embed_region_code_for_number" => [[P], P],
      "aether_pn_embed_national_significant_number" => [[P], P],
      "aether_pn_embed_length_of_ndc" => [[P], I],
      "aether_pn_embed_length_of_area_code" => [[P], I],
      "aether_pn_embed_is_geographical" => [[P], I],
      # validation
      "aether_pn_embed_is_possible_number" => [[P, P], I],
      "aether_pn_embed_is_possible_number_with_reason" => [[P, P], I],
      "aether_pn_embed_is_valid_number" => [[P, P], I],
      "aether_pn_embed_is_valid_number_for_region" => [[P, P], I],
      "aether_pn_embed_number_type" => [[P, P], I],
      "aether_pn_embed_can_be_internationally_dialled" => [[P, P], I],
      # formatting
      "aether_pn_embed_format" => [[P, P, I], P],
      "aether_pn_embed_format_out_of_country" => [[P, P, P], P],
      "aether_pn_embed_format_in_original" => [[P, P], P],
      # relations / helpers
      "aether_pn_embed_is_number_match" => [[P, P], I],
      "aether_pn_embed_truncate_too_long" => [[P, P], P],
      "aether_pn_embed_normalize_digits_only" => [[P], P],
      "aether_pn_embed_convert_alpha_characters" => [[P], P],
      "aether_pn_embed_is_alpha_number" => [[P], I],
      # AsYouType
      "aether_pn_embed_ayt_new" => [[P], P],
      "aether_pn_embed_ayt_input" => [[P, P], P],
      "aether_pn_embed_ayt_result" => [[P], P],
      "aether_pn_embed_ayt_clear" => [[P], P],
      # matcher
      "aether_pn_embed_matcher_count" => [[P, P, I], I],
      "aether_pn_embed_matcher_start" => [[P, P, I, I], I],
      "aether_pn_embed_matcher_end" => [[P, P, I, I], I],
      "aether_pn_embed_matcher_raw" => [[P, P, I, I], P],
      # ShortNumberInfo
      "aether_pn_embed_short_is_possible" => [[P, P], I],
      "aether_pn_embed_short_is_valid" => [[P, P], I],
      "aether_pn_embed_short_is_emergency" => [[P, P], I],
      "aether_pn_embed_short_connects_to_emergency" => [[P, P], I],
      "aether_pn_embed_short_is_carrier_specific" => [[P, P], I],
      "aether_pn_embed_short_is_sms_service" => [[P, P], I],
      "aether_pn_embed_short_expected_cost" => [[P, P], I],
      "aether_pn_embed_short_example_number" => [[P], P],
      # PhoneNumberToTimeZonesMapper
      "aether_pn_embed_tz_count" => [[P, P], I],
      "aether_pn_embed_tz_at" => [[P, P, I], P],
      "aether_pn_embed_tz_all" => [[P, P], P],
      "aether_pn_embed_tz_unknown" => [[], P],
      # PhoneNumberToCarrierMapper
      "aether_pn_embed_carrier_name" => [[P, P], P],
      "aether_pn_embed_carrier_name_for_valid" => [[P, P], P]
    }.freeze

    # A loaded engine: the Fiddle::Handle plus a memoized Fiddle::Function per
    # exported symbol. `fn.call("aether_pn_embed_country_code", ...)` is the
    # whole calling convention.
    class Lib
      attr_reader :path

      def initialize(path)
        @path = path
        @handle = Fiddle.dlopen(path)
        @fns = {}
        Native::SIGS.each do |name, (argtypes, restype)|
          @fns[name] = Fiddle::Function.new(@handle[name], argtypes, restype)
        rescue Fiddle::DLError => e
          raise Fiddle::DLError, "missing symbol #{name} in #{path}: #{e.message}"
        end
      end

      def call(name, *args)
        (@fns[name] || raise(ArgumentError, "unknown ABI symbol #{name}")).call(*args)
      end

      # Copy an ABI-returned string out and free it through the ABI.
      #
      # Every char* the engine returns is caller-owned; leaking it is the
      # single easiest mistake to make in any of these bindings, so all string
      # reads go through this one method.
      def take_string(ptr)
        return "" if ptr.nil?
        return "" if ptr.is_a?(Integer) && ptr.zero?

        p = ptr.is_a?(Fiddle::Pointer) ? ptr : Fiddle::Pointer.new(ptr)
        return "" if p.null?

        begin
          p.to_s.force_encoding(Encoding::UTF_8)
        ensure
          call("aether_pn_embed_free_string", p)
        end
      end
    end

    class << self
      # Load the engine .so, caching it process-wide. Returns a Lib.
      def load(path = nil)
        return @lib if @lib && path.nil?

        last = nil
        lib = nil
        candidates(path).each do |cand|
          lib = Lib.new(cand)
          break
        rescue Fiddle::DLError => e
          last = e
        end
        if lib.nil?
          raise Fiddle::DLError,
                "could not load the phonenumber engine (#{LIB_NAME}). Set " \
                "LIBPHONENUMBER_AE_LIB to its absolute path, or install a gem " \
                "that bundles it. Last error: #{last}"
        end

        @lib = lib if path.nil?
        lib
      end

      def candidates(explicit = nil)
        return [explicit] if explicit && !explicit.empty?

        out = []
        env = ENV["LIBPHONENUMBER_AE_LIB"]
        out << env if env && !env.empty?
        here = File.dirname(File.expand_path(__dir__)) # ruby/lib
        out << File.join(here, "..", "native", LIB_NAME)
        out << File.join(here, "phonenumber_ae", "native", LIB_NAME)
        out << LIB_NAME
        out
      end
    end
  end
end

File: javascript/aether/lib/native.js
'use strict';
/**
 * koffi bindings for the phonenumber engine (libphonenumber_ae.so), ABI v7.
 *
 * This module is the ONLY place in the JavaScript binding that knows about the
 * C ABI. Everything above it (`phonenumber.js`) is idiomatic JavaScript over
 * these symbols. No phone-number logic lives here or anywhere else in this
 * package — the engine is `core/phonenumber.ae`, shared by every binding.
 *
 * v7 makes the carrier + geocoder mappers multi-language: the four
 * PhoneNumberToCarrierMapper / PhoneNumberOfflineGeocoder symbols each gained a
 * trailing `const char *lang` (ISO code; "en" is always available and the
 * fallback). No new symbols — still 66. v6 added the PhoneNumberOfflineGeocoder
 * (2 symbols) on top of the v5 PhoneNumberToTimeZonesMapper (4 symbols) and
 * PhoneNumberToCarrierMapper (2 symbols) side-libraries and the v3
 * ShortNumberInfo ABI.
 * Every signature is still scalar-only
 * (`const char *` and `int`), and every returned `char*` is caller-owned. There
 * are still no opaque handles: a parsed number and an AsYouType state are
 * themselves caller-owned *strings* you hand back to the accessor calls and
 * free like any other returned string.
 *
 * Library resolution, in order:
 *   1. an explicit path passed to `load(path)`
 *   2. $LIBPHONENUMBER_AE_LIB          (what the in-tree .tests.ae leaf sets)
 *   3. native/ bundled next to this package (what an installed tarball ships)
 *   4. the OS loader's own search path
 */

const koffi = require('koffi');
const path = require('path');

const LIB_NAME = {
  darwin: 'libphonenumber_ae.dylib',
  win32: 'phonenumber_ae.dll',
}[process.platform] || 'libphonenumber_ae.so';

// ---- format styles (ABI constants — append only, never renumber) ----
// NOTE: v2 renumbered these. E164 is now 0 (was 2 in v1).
const E164 = 0;
const INTERNATIONAL = 1;
const NATIONAL = 2;
const RFC3966 = 3;

// ---- number types (number_type result; -1 = unknown) ----
const TYPE_UNKNOWN = -1;
const TYPE_FIXED_LINE = 0;
const TYPE_MOBILE = 1;
const TYPE_TOLL_FREE = 2;
const TYPE_PREMIUM_RATE = 3;
const TYPE_SHARED_COST = 4;
const TYPE_VOIP = 5;
const TYPE_PERSONAL_NUMBER = 6;
const TYPE_PAGER = 7;
const TYPE_UAN = 8;
const TYPE_VOICEMAIL = 9;
const TYPE_FIXED_LINE_OR_MOBILE = 10;

// ---- ValidationResult (is_possible_number_with_reason) ----
const VR_IS_POSSIBLE = 0;
const VR_IS_POSSIBLE_LOCAL_ONLY = 4;
const VR_INVALID_COUNTRY_CODE = 1;
const VR_TOO_SHORT = 2;
const VR_INVALID_LENGTH = 5;
const VR_TOO_LONG = 3;

// ---- MatchType (is_number_match) ----
const MATCH_NOT_A_NUMBER = 0;
const MATCH_NO_MATCH = 1;
const MATCH_SHORT_NSN = 2;
const MATCH_NSN = 3;
const MATCH_EXACT = 4;

// ---- CountryCodeSource (pn_source) ----
const SRC_FROM_NUMBER_WITH_PLUS = 1;
const SRC_FROM_NUMBER_WITH_IDD = 5;
const SRC_FROM_NUMBER_WITHOUT_PLUS = 10;
const SRC_FROM_DEFAULT_COUNTRY = 20;

// ---- matcher leniency ----
const LENIENCY_POSSIBLE = 0;
const LENIENCY_VALID = 1;

// ---- ShortNumberCost (short_expected_cost) ----
const COST_TOLL_FREE = 0;
const COST_STANDARD_RATE = 1;
const COST_PREMIUM_RATE = 2;
const COST_UNKNOWN = 3;

let cached = null;

function* candidates(explicit) {
  if (explicit) {
    yield explicit;
    return;
  }
  const env = process.env.LIBPHONENUMBER_AE_LIB;
  if (env) yield env;
  yield path.join(__dirname, '..', 'native', LIB_NAME);
  yield LIB_NAME;
}

/**
 * Load the engine .so, caching it process-wide. Returns the symbol table.
 */
function load(explicit) {
  if (cached !== null && !explicit) return cached;

  let lib = null;
  let last = null;
  for (const cand of candidates(explicit)) {
    try {
      lib = koffi.load(cand);
      break;
    } catch (err) {
      last = err;
    }
  }
  if (lib === null) {
    throw new Error(
      `could not load the phonenumber engine (${LIB_NAME}). Set ` +
      'LIBPHONENUMBER_AE_LIB to its absolute path, or install a package that ' +
      `bundles it. Last error: ${last && last.message}`);
  }

  const api = declare(lib);
  if (!explicit) cached = api;
  return api;
}

/**
 * Bind every exported symbol. Signature shapes come straight from
 * core/embed.ae; the mangled `aether_pn_embed_` prefix is what `--emit=lib`
 * produces.
 *
 * Note the return type of every string-returning ABI call is `void *`, not
 * `const char *`: koffi would otherwise decode-and-forget the pointer and we
 * could never hand it back to aether_pn_embed_free_string. Every one of those
 * is caller-owned, and leaking it is the single easiest mistake in a binding.
 */
function declare(lib) {
  const f = (sig) => lib.func(sig);
  return {
    _lib: lib,

    abiVersion: f('int aether_pn_embed_abi_version()'),
    freeString: f('void aether_pn_embed_free_string(void *s)'),

    // ---- metadata ----
    countryCode: f('void *aether_pn_embed_country_code(const char *region)'),
    exampleNumber: f('void *aether_pn_embed_example_number(const char *region)'),
    exampleNumberForType: f('void *aether_pn_embed_example_number_for_type(const char *region, int type)'),
    invalidExampleNumber: f('void *aether_pn_embed_invalid_example_number(const char *region)'),
    possibleLengths: f('void *aether_pn_embed_possible_lengths(const char *region)'),
    regionCodeForCountryCode: f('void *aether_pn_embed_region_code_for_country_code(const char *cc)'),
    isNanpaCountry: f('int aether_pn_embed_is_nanpa_country(const char *region)'),
    nddPrefixForRegion: f('void *aether_pn_embed_ndd_prefix_for_region(const char *region, int stripNonDigits)'),
    regionCount: f('int aether_pn_embed_region_count()'),
    regionAt: f('void *aether_pn_embed_region_at(int index)'),
    ccRegionCount: f('int aether_pn_embed_cc_region_count(const char *cc)'),
    ccRegionAt: f('void *aether_pn_embed_cc_region_at(const char *cc, int index)'),

    // ---- parse + parsed-number accessors ----
    parse: f('void *aether_pn_embed_parse(const char *input, const char *region)'),
    nationalNumber: f('void *aether_pn_embed_national_number(const char *region, const char *input)'),
    pnRegion: f('void *aether_pn_embed_pn_region(const char *pn)'),
    pnCountryCode: f('void *aether_pn_embed_pn_country_code(const char *pn)'),
    pnNationalNumber: f('void *aether_pn_embed_pn_national_number(const char *pn)'),
    pnExtension: f('void *aether_pn_embed_pn_extension(const char *pn)'),
    pnItalianLeadingZero: f('int aether_pn_embed_pn_italian_leading_zero(const char *pn)'),
    pnSource: f('int aether_pn_embed_pn_source(const char *pn)'),
    pnError: f('void *aether_pn_embed_pn_error(const char *pn)'),
    regionCodeForNumber: f('void *aether_pn_embed_region_code_for_number(const char *pn)'),
    nationalSignificantNumber: f('void *aether_pn_embed_national_significant_number(const char *pn)'),
    lengthOfNdc: f('int aether_pn_embed_length_of_ndc(const char *pn)'),
    lengthOfAreaCode: f('int aether_pn_embed_length_of_area_code(const char *pn)'),
    isGeographical: f('int aether_pn_embed_is_geographical(const char *pn)'),

    // ---- validation ----
    isPossibleNumber: f('int aether_pn_embed_is_possible_number(const char *region, const char *input)'),
    isPossibleNumberWithReason: f('int aether_pn_embed_is_possible_number_with_reason(const char *region, const char *input)'),
    isValidNumber: f('int aether_pn_embed_is_valid_number(const char *region, const char *input)'),
    isValidNumberForRegion: f('int aether_pn_embed_is_valid_number_for_region(const char *input, const char *region)'),
    numberType: f('int aether_pn_embed_number_type(const char *region, const char *input)'),
    canBeInternationallyDialled: f('int aether_pn_embed_can_be_internationally_dialled(const char *region, const char *input)'),

    // ---- formatting ----
    format: f('void *aether_pn_embed_format(const char *region, const char *input, int fmt)'),
    formatOutOfCountry: f('void *aether_pn_embed_format_out_of_country(const char *region, const char *input, const char *callingFrom)'),
    formatInOriginal: f('void *aether_pn_embed_format_in_original(const char *pn, const char *callingFrom)'),

    // ---- relations / helpers ----
    isNumberMatch: f('int aether_pn_embed_is_number_match(const char *a, const char *b)'),
    truncateTooLong: f('void *aether_pn_embed_truncate_too_long(const char *region, const char *input)'),
    normalizeDigitsOnly: f('void *aether_pn_embed_normalize_digits_only(const char *s)'),
    convertAlphaCharacters: f('void *aether_pn_embed_convert_alpha_characters(const char *s)'),
    isAlphaNumber: f('int aether_pn_embed_is_alpha_number(const char *s)'),

    // ---- AsYouTypeFormatter (state threaded as a caller-owned string) ----
    aytNew: f('void *aether_pn_embed_ayt_new(const char *region)'),
    aytInput: f('void *aether_pn_embed_ayt_input(const char *state, const char *ch)'),
    aytResult: f('void *aether_pn_embed_ayt_result(const char *state)'),
    aytClear: f('void *aether_pn_embed_ayt_clear(const char *state)'),

    // ---- PhoneNumberMatcher / findNumbers ----
    matcherCount: f('int aether_pn_embed_matcher_count(const char *text, const char *region, int leniency)'),
    matcherStart: f('int aether_pn_embed_matcher_start(const char *text, const char *region, int leniency, int idx)'),
    matcherEnd: f('int aether_pn_embed_matcher_end(const char *text, const char *region, int leniency, int idx)'),
    matcherRaw: f('void *aether_pn_embed_matcher_raw(const char *text, const char *region, int leniency, int idx)'),

    // ---- ShortNumberInfo (short / emergency numbers) ----
    shortIsPossible: f('int aether_pn_embed_short_is_possible(const char *region, const char *input)'),
    shortIsValid: f('int aether_pn_embed_short_is_valid(const char *region, const char *input)'),
    shortIsEmergency: f('int aether_pn_embed_short_is_emergency(const char *region, const char *input)'),
    shortConnectsToEmergency: f('int aether_pn_embed_short_connects_to_emergency(const char *region, const char *input)'),
    shortIsCarrierSpecific: f('int aether_pn_embed_short_is_carrier_specific(const char *region, const char *input)'),
    shortIsSmsService: f('int aether_pn_embed_short_is_sms_service(const char *region, const char *input)'),
    shortExpectedCost: f('int aether_pn_embed_short_expected_cost(const char *region, const char *input)'),
    shortExampleNumber: f('void *aether_pn_embed_short_example_number(const char *region)'),

    // ---- PhoneNumberToTimeZonesMapper (timezone lookup) ----
    tzCount: f('int aether_pn_embed_tz_count(const char *region, const char *input)'),
    tzAt: f('void *aether_pn_embed_tz_at(const char *region, const char *input, int idx)'),
    tzAll: f('void *aether_pn_embed_tz_all(const char *region, const char *input)'),
    tzUnknown: f('void *aether_pn_embed_tz_unknown()'),

    // ---- PhoneNumberToCarrierMapper (localized carrier names) ----
    carrierName: f('void *aether_pn_embed_carrier_name(const char *region, const char *input, const char *lang)'),
    carrierNameForValid: f('void *aether_pn_embed_carrier_name_for_valid(const char *region, const char *input, const char *lang)'),

    // ---- PhoneNumberOfflineGeocoder (localized geographic descriptions) ----
    geoDescription: f('void *aether_pn_embed_geo_description(const char *region, const char *input, const char *lang)'),
    geoDescriptionForValid: f('void *aether_pn_embed_geo_description_for_valid(const char *region, const char *input, const char *lang)'),
  };
}

/**
 * Copy an ABI-returned string out and free it through the ABI.
 *
 * Every char* the engine returns is caller-owned; leaking it is the single
 * easiest mistake to make in any of these bindings.
 *
 * Koffi 3 represents pointers as BigInt, and a null pointer as `null` — so the
 * falsy check covers both `null` and `0n`.
 */
function takeString(api, ptr) {
  if (!ptr) return '';
  try {
    return koffi.decode.string(ptr);
  } finally {
    api.freeString(ptr);
  }
}

module.exports = {
  load,
  takeString,
  koffi,
  LIB_NAME,
  E164, INTERNATIONAL, NATIONAL, RFC3966,
  TYPE_UNKNOWN, TYPE_FIXED_LINE, TYPE_MOBILE, TYPE_TOLL_FREE,
  TYPE_PREMIUM_RATE, TYPE_SHARED_COST, TYPE_VOIP, TYPE_PERSONAL_NUMBER,
  TYPE_PAGER, TYPE_UAN, TYPE_VOICEMAIL, TYPE_FIXED_LINE_OR_MOBILE,
  VR_IS_POSSIBLE, VR_IS_POSSIBLE_LOCAL_ONLY, VR_INVALID_COUNTRY_CODE,
  VR_TOO_SHORT, VR_INVALID_LENGTH, VR_TOO_LONG,
  MATCH_NOT_A_NUMBER, MATCH_NO_MATCH, MATCH_SHORT_NSN, MATCH_NSN, MATCH_EXACT,
  SRC_FROM_NUMBER_WITH_PLUS, SRC_FROM_NUMBER_WITH_IDD,
  SRC_FROM_NUMBER_WITHOUT_PLUS, SRC_FROM_DEFAULT_COUNTRY,
  LENIENCY_POSSIBLE, LENIENCY_VALID,
  COST_TOLL_FREE, COST_STANDARD_RATE, COST_PREMIUM_RATE, COST_UNKNOWN,
};
